src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,46 +1,100 @@
 package com.es.phoneshop.model.product;
 
-import java.math.BigDecimal;
-import java.util.Currency;
+import com.es.phoneshop.model.exception.ProductNotFoundException;
+
+import java.util.Comparator;
 import java.util.List;
-import java.util.NoSuchElementException;
+import java.util.Map;
+import java.util.TreeMap;
 import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private static ProductDao productDao;
+
+    public static ProductDao getInstance(){
+        if (productDao==null){
+            synchronized (ArrayListProductDao.class){
+                if (productDao==null)
+                    productDao=new ArrayListProductDao();
+            }
```

Comment:
Here and in other places: consider using curly braces by default since this improves code readability and maintainability.

Compare:
```
if (productDao==null)
    productDao=new ArrayListProductDao();
```

and:

```
if (productDao==null) {
    productDao=new ArrayListProductDao();
}
```

If we need to add additional line in if block it's MUCH more chances to make a mistake like:
```
if (productDao==null)
    productDao=new ArrayListProductDao();
    ...some code here...
```
If this case ...some code here... will not be executed in if block.